src/web-worker/amp-worker.js

```diff
@@ -0,0 +1,145 @@
+ * limitations under the License.
+ */
+
+import {FromWorkerMessageDef, ToWorkerMessageDef} from './web-worker-defines';
+import {calculateEntryPointScriptUrl} from '../service/extension-location';
+import {dev} from '../log';
+import {fromClass} from '../service';
+import {isExperimentOn} from '../experiments';
+import {getMode} from '../mode';
+
+const TAG = 'web-worker';
+
+/**
+ * @typedef {{method: string, resolve: !Function, reject: !Function}}
+ */
+let PendingMessageDef;
+
+/**
+ * Invokes function named `method` with args `opt_args` on the web worker
+ * and returns a Promise that will be resolved with the function's return value.
+ * @note Currently only works in a single entry point.
+ * @param {!Window} win
+ * @param {string} method
+ * @param {!Array=} opt_args
+ * @return {!Promise}
+ */
+export function invokeWebWorker(win, method, opt_args) {
+  if (!isExperimentOn(win, TAG)) {
+    return Promise.reject(`Experiment "${TAG}" is disabled.`);
+  }
+  if (!win.Worker) {
+    return Promise.reject('Worker not supported in window: ' + win);
+  }
+  const worker = fromClass(win, 'amp-worker', AmpWorker);
+  return worker.sendMessage_(method, opt_args || []);
+}
+
+/**
+ * @param {!Window} win
+ * @return {!AmpWorker}
+ * @visibleForTesting
+ */
+export function ampWorkerForTesting(win) {
+  return fromClass(win, 'amp-worker', AmpWorker);
+}
+
+/**
+ * A Promise-based API wrapper around a single Web Worker.
+ * @private
+ */
+class AmpWorker {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const @private {!Window} */
+    this.win_ = win;
+
+    const url =
+        calculateEntryPointScriptUrl(location, 'ww', getMode().localDev);
+    /** @const @private {!Worker} */
+    this.worker_ = new win.Worker(url);
+    this.worker_.onmessage = this.receiveMessage_.bind(this);
+
+    /**
+     * Array of in-flight messages pending response from worker.
+     * @private {!Array<(PendingMessageDef|undefined)>}
+     */
+    this.messages_ = [];
+  }
+
+  /**
+   * Sends a method invocation request to the worker and returns a Promise.
+   * @param {string} method
+   * @param {!Array} args
+   * @return {!Promise}
+   * @private
+   */
+  sendMessage_(method, args) {
+    const promise = new Promise((resolve, reject) => {
+      const index = this.messages_.length;
+      this.messages_[index] = {method, resolve, reject};
+
+      /** @type {ToWorkerMessageDef} */
+      const message = {method, args, id: index};
+      this.worker_./*OK*/postMessage(message);
+    });
+    return promise;
+  }
+
+  /**
+   * Receives the result of a method invocation from the worker and resolves
+   * the Promise returned from the corresponding `sendMessage_()` call.
+   * @param {!MessageEvent} event
+   * @private
+   */
+  receiveMessage_(event) {
+    const {method, returnValue, id} =
+        /** @type {FromWorkerMessageDef} */ (event.data);
+
+    const message = this.messages_[id];
```

Comment:
Only while messages are in flight which shouldn't be an issue for current functionality.